ONSdigital/dp-kafka
# refactored healthcheck to set status to WARNING if at least one broke…

### What

Kafka health check to display `WARNING` if enough brokers are healthy, but not all of them. Trello card: https://trello.com/c/NmmQVJGg/5435-kafka-client-fails-health-check-when-msk-security-patches-are-applied-s1

- Refactored kafka healthcheck code to allow more flexibility (and introduced HealthInfo to make it clearer)
- Consumers are healthy if at least one broker is healthy
- Producers are healthy if at least two brokers are healthy
- Healthcheck no longer tries to initialise the producer/consumer (this is also done in a retry loop, so it is unnecessary and may be confusing)

Describe what you have changed and why.

### How to review

- Make sure code changes make sense
- Make sure unit tests pass

- (not required, I tested this already ) - Validate examples work as expected:
  - checkout `dp-compose` with kafka cluster https://github.com/ONSdigital/dp-kafka/pull/46
  - run `make start` in `dp-compose`
  - run the kafka examples and validate that they produce and consume messages as expected
  - kill one broker container
  - validate that the kafka containers can still produce and consume messages.
 

- (not required, I tested this already) - Validate that services show healthcheck as expected:
  - checkout `dp-compose` with kafka cluster https://github.com/ONSdigital/dp-kafka/pull/46
  - run `make start` in `dp-compose`
  - use the latest commit of this PR in a service that has kafka producers and consumers (e.g. dataset-exporter): `go get github.com/ONSdigital/dp-kafka/v2@655f71b035d2ddf19a7bdfad915ff15c9fcfba85`
  - Modify the config of the service to use the 3 brokers of the cluster: in `config.go` set `KafkaAddr:                  []string{"localhost:19092", "localhost:29092", "localhost:39092"}`
  - run the service: `make debug`
  - check that once the system is ready, `/health` displays 'OK' messages for the kafka consumers and procuders
  - kill one broker container
  - check that `/health` displays `WARNING` messages for kafka consumers and producers
  - kill another broker container
  - check that `/health` displays `WARNING` messages for kafka consumers and `CRITICAL` for producers

### Who can review

Anyone
